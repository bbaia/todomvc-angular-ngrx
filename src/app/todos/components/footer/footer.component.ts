import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TodoFilter } from '../../models';

@Component({
  selector: 'app-footer',
  templateUrl: './footer.component.html',
})
export class FooterComponent {
  @Input() hasCompletedTodos: boolean | null = null;
  @Input() undoneTodosCount: number | null = null;
  @Input() currentFilter: TodoFilter | null = null;
  @Output() filter = new EventEmitter<TodoFilter>();
  @Output() clearCompleted = new EventEmitter();
}
