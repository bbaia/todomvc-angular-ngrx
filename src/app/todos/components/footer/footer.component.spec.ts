import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { TodoFilter } from '../../models';
import { FooterComponent } from './footer.component';

describe('FooterComponent', () => {
  let component: FooterComponent;
  let fixture: ComponentFixture<FooterComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule.withRoutes([])],
      declarations: [FooterComponent],
    });
    fixture = TestBed.createComponent(FooterComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeDefined();
  });

  it('should render', () => {
    component.hasCompletedTodos = true;
    component.undoneTodosCount = 1;
    component.currentFilter = 'SHOW_ALL';
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should raise "filter" event', done => {
    component.hasCompletedTodos = false;
    component.undoneTodosCount = 1;
    component.currentFilter = 'SHOW_ALL';
    fixture.detectChanges();

    component.filter.subscribe((filter: TodoFilter) => {
      expect(filter).toBe('SHOW_ACTIVE');
      done();
    });
    fixture.debugElement.nativeElement
      .querySelector('.filters > li:nth-of-type(2) > a')
      .click();
  });

  it('should raise "clearCompleted" event', done => {
    component.hasCompletedTodos = true;
    component.undoneTodosCount = 1;
    component.currentFilter = 'SHOW_ALL';
    fixture.detectChanges();

    component.clearCompleted.subscribe(() => {
      done();
    });
    fixture.debugElement.nativeElement
      .querySelector('.clear-completed')
      .click();
  });

  it('should hide "Clear completed" button with nothing completed', () => {
    component.hasCompletedTodos = false;
    component.undoneTodosCount = 1;
    component.currentFilter = 'SHOW_ALL';
    fixture.detectChanges();
    expect(
      fixture.debugElement.nativeElement.querySelector('.clear-completed'),
    ).toBeNull();
  });

  it('should display undone todos', () => {
    component.hasCompletedTodos = false;
    component.undoneTodosCount = 1;
    component.currentFilter = 'SHOW_ALL';
    fixture.detectChanges();
    expect(
      fixture.debugElement.nativeElement.querySelector('.todo-count')
        .textContent,
    ).toEqual(`${component.undoneTodosCount} items left`);
  });

  it('should select current filter', () => {
    component.hasCompletedTodos = false;
    component.undoneTodosCount = 1;
    component.currentFilter = 'SHOW_ACTIVE';
    fixture.detectChanges();
    expect(
      fixture.debugElement.nativeElement
        .querySelector('.filters > li > a.selected')
        .textContent.trim(),
    ).toEqual(`Active`);
  });
});
