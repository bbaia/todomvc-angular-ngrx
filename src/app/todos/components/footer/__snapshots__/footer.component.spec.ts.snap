// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FooterComponent should render 1`] = `
<app-footer
  __ngContext__={[Function LRootView]}
  clearCompleted={[Function EventEmitter_]}
  currentFilter={[Function String]}
  filter={[Function EventEmitter_]}
  hasCompletedTodos={[Function Boolean]}
  undoneTodosCount={[Function Number]}
>
  <footer
    class="footer"
    id="footer"
  >
    <span
      class="todo-count"
      id="todo-count"
    >
      1 items left
    </span>
    <ul
      class="filters"
      id="filters"
    >
      <li>
        <a
          class="selected"
          href="/"
        >
           All 
        </a>
      </li>
      <li>
        <a
          href="/"
        >
           Active 
        </a>
      </li>
      <li>
        <a
          href="/"
        >
           Completed 
        </a>
      </li>
    </ul>
    <button
      class="clear-completed"
      id="clear-completed"
    >
       Clear completed 
    </button>
  </footer>
</app-footer>
`;
